api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java

```diff
@@ -425,85 +431,177 @@ public PatientIdentifierType getPatientIdentifierType(Integer patientIdentifierT
-			for (Field f : patient.getDeclaredFields()) {
-				patientFieldNames.add(f.getName());
-				log.debug(f.getName());
-			}
-			
-			Class person = Person.class;
-			Set<String> personFieldNames = new HashSet<String>(person.getDeclaredFields().length);
-			for (Field f : person.getDeclaredFields()) {
-				personFieldNames.add(f.getName());
-				log.debug(f.getName());
+
+			String sqlString = getDuplicatePatientsSQLString(attributes);
+
+			try {
+				SQLQuery sqlquery = sessionFactory.getCurrentSession().createSQLQuery(sqlString);
+				patientIds = sqlquery.list();
+				if (patientIds.size() >= 1) {
+					Query query = sessionFactory.getCurrentSession().createQuery(
+							"from Patient p1 where p1.patientId in (:ids)");
+					query.setParameterList("ids", patientIds);
+					patients = query.list();
+				}
 			}
-			
-			Class personName = PersonName.class;
-			Set<String> personNameFieldNames = new HashSet<String>(personName.getDeclaredFields().length);
-			for (Field f : personName.getDeclaredFields()) {
-				personNameFieldNames.add(f.getName());
-				log.debug(f.getName());
+			catch (HibernateException ex) {
+				log.warn("Hibernate exception: " + ex.getStackTrace().toString());
+				throw ex;
+
 			}
-			
-			Class identifier = PatientIdentifier.class;
-			Set<String> identifierFieldNames = new HashSet<String>(identifier.getDeclaredFields().length);
-			for (Field f : identifier.getDeclaredFields()) {
-				identifierFieldNames.add(f.getName());
-				log.debug(f.getName());
+			finally {}
+		}
+
+		sortDuplicatePatients(patients, patientIds);
+		return patients;
+	}
+
+	private String getDuplicatePatientsSQLString(List<String> attributes) {
+		String outerSelect = "select distinct t1.patient_id from patient t1 ";
+
+		Class patient = Patient.class;
+		Set<String> patientFieldNames = new HashSet<String>(patient.getDeclaredFields().length);
+		for (Field f : patient.getDeclaredFields()) {
+			patientFieldNames.add(f.getName());
+			log.debug(f.getName());
+		}
+
+		Class person = Person.class;
+		Set<String> personFieldNames = new HashSet<String>(person.getDeclaredFields().length);
+		for (Field f : person.getDeclaredFields()) {
+			personFieldNames.add(f.getName());
+			log.debug(f.getName());
+		}
+
+		Class personName = PersonName.class;
+
+		Set<String> personNameFieldNames = new HashSet<String>(personName.getDeclaredFields().length);
+		for (Field f : personName.getDeclaredFields()) {
+			personNameFieldNames.add(f.getName());
+			log.debug(f.getName());
+		}
+
+		Class identifier = PatientIdentifier.class;
+		Set<String> identifierFieldNames = new HashSet<String>(identifier.getDeclaredFields().length);
+		for (Field f : identifier.getDeclaredFields()) {
+			identifierFieldNames.add(f.getName());
+			log.debug(f.getName());
+		}
+		List<String> whereConditions = new ArrayList<String>();
+
+
+		List<String> innerFields = new ArrayList<String>();
+		String innerSelect = " from patient p1 ";
+
+		for (String s : attributes) {
```

Comment:
@karrtikiyer-tw can you please use more expressive variable names, like maybe `attribute` in this case. your method is very long so one has to mentally keep track of all the different variables. good names make it easier, thanks :smile: